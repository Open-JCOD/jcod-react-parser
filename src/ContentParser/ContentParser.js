import React, { Fragment, PureComponent } from 'react'
import PropTypes from 'prop-types'

export const stringerJoin = arrOfString => {
  // return arrOfString.join(', ')
  return arrOfString.reduce((str, entry, index, arr) => {
    if (index === 0) return str + entry
    if (index === arr.length - 1) return `${str} and ${entry}`
    return `${str}, ${entry}`
  }, '')
}

const componentType = PropTypes.oneOfType([
  PropTypes.string,
  PropTypes.shape({
    key: PropTypes.string,
    component: PropTypes.string.isRequired,
    props: PropTypes.object,
    renderProps: PropTypes.object,
  }),
])

const defaultSpreader = {
  formater: (child, metadata) => child,
  nodeDecorator: (node, metadata) => node,
  treeDecorator: (tree, metadata) => tree,
  // subTreeDecorator: (tree, metadata) => tree,
}

export default class ContentParser extends PureComponent {
  static propTypes = {
    data: PropTypes.arrayOf(componentType).isRequired,
    components: PropTypes.object,
    options: PropTypes.shape({
      allowUnsecureElements: PropTypes.oneOfType([
        PropTypes.bool,
        PropTypes.array,
      ]),
      allowUnsecureHtmlElement: PropTypes.bool,
      allowUnsecureCustomElement: PropTypes.bool,
      displayErrorMessage: PropTypes.bool,
      __DEPRECATED__renderAllChildren: PropTypes.bool,
    }),
    spreader: PropTypes.shape({
      formater: PropTypes.func,
      nodeDecorator: PropTypes.func,
      treeDecorator: PropTypes.func,
      // subTreeDecorator: PropTypes.func,
    }),
  }

  static defaultProps = {
    data: [],
    components: {},
    options: {
      allowUnsecureHtmlElement: false,
      allowUnsecureCustomElement: false, // http://w3c.github.io/webcomponents/spec/custom/#valid-custom-element-name
      allowUnsecureElements: false,
      displayErrorMessage: true,
      __DEPRECATED__renderAllChildren: false,
    },
    spreader: defaultSpreader,
  }

  static getTypeOfComponent = component => {
    return {
      isHtmlElement: component.search(/(?:[A-Z]+)|(?:-+)/) < 0,
      isCustomElement: component.search(/-+/) >= 0,
    }
  }

  static FalsyComponent = function FalsyComponent() {
    return null
  }
  static getFalsyComponent = (
    component,
    typeOfComponent,
    displayErrorMessage,
  ) => {
    const errorMessage = [].concat(
      `[${component}] can not be rendered: \n  it is not an available component.`,
      typeOfComponent.isHtmlElement
        ? `\n> [${component}] seems to be a [HTML Element <${component}>]. \n  Maybe should you set the 'allowUnsecureHtmlElement' option to true ?`
        : [],
      typeOfComponent.isCustomElement
        ? `\n> [${component}] seems to be a [Custom Element <${component}>]. \n  Maybe should you set the 'allowUnsecureCustomElement' option to true ?`
        : [],
    )
    displayErrorMessage && console.error && console.error(...errorMessage)
    return ContentParser.FalsyComponent
  }

  static showError = (jcodNode, metadata = {}) => {
    if (!jcodNode || typeof jcodNode !== 'object') return
    const { path: jcodPath } = metadata
    const {
      component,
      // key,
      children,
      props: { key: jcodPropsKey, children: jcodPropsChildren } = {},
      renderProps: {
        key: jcodRenderPropsKey,
        children: jcodRenderPropsChildren,
      } = {},
    } = jcodNode
    const childrenValues = [].concat(
      children ? '[ROOT]' : [],
      jcodRenderPropsChildren ? '[renderProps]' : [],
      jcodPropsChildren ? '[props]' : [],
    )
    const errorMessage = [].concat(
      childrenValues.length > 1
        ? `The children of component [${component} // in position ${jcodPath}] have several values defined in ${stringerJoin(
            childrenValues,
          )}. The value of the children used is the one defined in ${
            childrenValues[0]
          }. Other values are ignored.`
        : [],
      jcodPropsKey
        ? `The [props.key] with the value '${jcodPropsKey}' of component [${component} // in position ${jcodPath}] is ignored. The key value, when defined, must be set in the ROOT of JCOD node.`
        : [],
      jcodRenderPropsKey
        ? `The [renderProps.key] with the value '${jcodRenderPropsKey}' of component [${component} // in position ${jcodPath}] is ignored. The key value, when defined, must be set in the ROOT of JCOD node.`
        : [],
    )
    errorMessage.length && console.error && console.error(...errorMessage)
  }

  static getComponent = (component, availableComponent = {}, options = {}) => {
    const typeOfComponent = ContentParser.getTypeOfComponent(component)

    if (options.allowUnsecureElements === true) {
      options.allowUnsecureCustomElement = true
      options.allowUnsecureHtmlElement = true
    }
    if (!Array.isArray(options.allowUnsecureElements))
      options.allowUnsecureElements = []

    return (
      availableComponent[component] ||
      (options.allowUnsecureElements.includes(component) && component) ||
      (options.allowUnsecureCustomElement &&
        typeOfComponent.isCustomElement &&
        component) ||
      (options.allowUnsecureHtmlElement &&
        typeOfComponent.isHtmlElement &&
        component) ||
      ContentParser.getFalsyComponent(component, typeOfComponent)
    )
  }

  constructor(props) {
    super(props)

    const { components, options: instanceOptions, spreader } = props
    const options = {
      ...ContentParser.defaultProps.options,
      ...instanceOptions,
    }

    const getRenderChild = (_spreader = {}) => (availableComp = {}) =>
      function renderChild(_path) {
        return (child, index, data) => {
          const jcodPath = `${_path ? `${_path}.` : ''}${index}`
          const spreader = {
            ...defaultSpreader,
            ..._spreader,
          }
          const getTranslateToInstance = index => entry => {
            if (typeof entry === 'undefined') return null
            return typeof entry === 'object'
              ? [].concat(entry).map(renderChild(jcodPath))
              : renderChild(jcodPath)(entry, index)
          }

          if (typeof child === 'string') {
            // TODO : Update formater for support string
            // const metadata = {
            //   path: jcodPath,
            //   jcod: { component: '[Text Node]', value: child },
            // }
            // return spreader.formater
            //   ? spreader.formater(child, metadata)
            //   : child
            return child
          }
          if (typeof child === 'object') {
            const metadata = {
              path: jcodPath,
              jcod: child,
            }
            const formatedJCOD = spreader.formater
              ? spreader.formater(child, metadata)
              : child
            const {
              component,
              // TODO : Get better system for get Key
              key = JSON.stringify(child),
              children: jcodChildren,
              props: jcodProps = {},
              renderProps: jcodRenderProps = {},
            } = formatedJCOD

            if (options.displayErrorMessage)
              ContentParser.showError(formatedJCOD, metadata)

            const Component = ContentParser.getComponent(
              component,
              availableComp,
              options,
            )

            const {
              key: jcodPropsKey,
              children: childrenProps,
              ...otherProps
            } = jcodProps
            const {
              key: jcodRenderPropsKey,
              children: childrenRenderProps,
              ...otherRenderProps
            } = jcodRenderProps

            const translateToInstances = getTranslateToInstance(index)

            const renderPropsComponents = Object.entries(
              otherRenderProps,
            ).reduce(
              (acc, [key, value]) => ({
                ...acc,
                [key]: translateToInstances(value),
              }),
              {},
            )

            const childsComponent =
              jcodChildren ||
              childrenRenderProps ||
              options.__DEPRECATED__renderAllChildren
                ? translateToInstances(
                    jcodChildren ||
                      childrenRenderProps ||
                      (options.__DEPRECATED__renderAllChildren &&
                        childrenProps),
                  )
                : childrenProps && childrenProps.toString()

            const instance = (
              <Component key={key} {...renderPropsComponents} {...otherProps}>
                {childsComponent}
              </Component>
            )

            return spreader.nodeDecorator
              ? spreader.nodeDecorator(instance)
              : instance
          }

          return child
        }
      }

    // ---------------------
    // ---------------------
    // ---------------------

    // TODO : Add support if 'components' is type Map
    const translatedComponents = Object.fromEntries
      ? Object.fromEntries(Object.entries(components))
      : Object.entries(components).reduce((acc, [componentName, component]) => {
          return {
            ...acc,
            [componentName]: component,
          }
        }, {})

    this.renderChild = getRenderChild(spreader)(translatedComponents)
  }

  render() {
    const { data, spreader = {} } = this.props
    const arrayData = Array.isArray(data) ? data : [data]

    return spreader.treeDecorator ? (
      spreader.treeDecorator(
        <Fragment>{arrayData.map(this.renderChild())}</Fragment>,
      )
    ) : (
      <Fragment>{arrayData.map(this.renderChild())}</Fragment>
    )
  }
}
