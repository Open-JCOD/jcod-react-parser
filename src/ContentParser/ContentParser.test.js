import React from 'react'
import renderer from 'react-test-renderer'
import mockConsole from 'jest-mock-console'

import ContentParser, { stringerJoin } from './ContentParser'

const typeComponent = {
  html: {
    isHtmlElement: true,
    isCustomElement: false,
  },
  customElement: {
    isHtmlElement: false,
    isCustomElement: true,
  },
  reactElement: {
    isHtmlElement: false,
    isCustomElement: false,
  },
}

test('adds 1 + 2 to equal 3', () => {
  expect(1 + 2).toBe(3)
})

describe('[stringerJoin] Test Helper to join literally an array of string', () => {
  const strArr = ['first', 'second', 'third', 'fourth']
  test('Should join Four items', () => {
    expect(stringerJoin(strArr)).toBe('first, second, third and fourth')
  })
  test('Should join three items', () => {
    expect(stringerJoin(strArr.slice(0, 3))).toBe('first, second and third')
  })
  test('Should join two items', () => {
    expect(stringerJoin(strArr.slice(0, 1))).toBe('first')
  })
  test('Should return one items', () => {
    expect(stringerJoin(strArr.slice(0, 1))).toBe('first')
  })
})

describe('[ContentParser.getTypeOfComponent] Identify type of Component.', () => {
  const tester = ContentParser.getTypeOfComponent
  test('Should identify HTML element.', () => {
    const testHtmlElement = tester('html')
    expect(testHtmlElement).toEqual(typeComponent.html)
  })
  test('Should identify Custom-Element', () => {
    const testCustomElement = tester('custom-element')
    expect(testCustomElement).toEqual(typeComponent.customElement)
  })
  test('Should identify React Component', () => {
    const testReactComponent = tester('ReactComponent')
    expect(testReactComponent).toEqual(typeComponent.reactElement)
  })
})

describe('[ContentParser.getFalsyComponent] Identify the not supported component.', () => {
  const restoreConsole = mockConsole()
  const element = 'react'

  const falsyComponent = ContentParser.getFalsyComponent(
    element,
    typeComponent.html,
    true,
  )
  const spyError = console.error.mock.calls

  test('Should return a function.', () => {
    expect(typeof falsyComponent).toBe('function')
  })

  test('This function should return null.', () => {
    const falsyComponentValue = falsyComponent()
    expect(falsyComponentValue).toBeNull()
  })

  test('Should display 1 error.', () => {
    mockConsole()
    // const spyError = console.error.mock.calls
    expect(spyError.length).toBe(1)
    // expect(spyError).toEqual([])
    // expect(console.error.mock.calls).toEqual([])
  })

  restoreConsole()
})

describe('[ContentParser.showError] Analyse JCOD node & send error.', () => {
  const metadata = { path: '0.0.1' }

  describe('Test children value', () => {
    test('Should display 1 error: Children in ROOT & Props.', () => {
      mockConsole()

      const jcodNode = {
        key: 'any key',
        component: 'React',
        children: 'something',
        props: {
          children: 'other something',
          className: 'any-class',
        },
      }
      const response = `The children of component [${jcodNode.component} // in position ${metadata.path}] have several values defined in [ROOT] and [props]. The value of the children used is the one defined in [ROOT]. Other values are ignored.`

      ContentParser.showError(jcodNode, metadata)
      const spyError = console.error.mock.calls

      expect(Array.isArray(spyError)).toBe(true)
      expect(spyError).toEqual([[response]])
    })

    test('Should display 1 error: Children in ROOT & RenderProps.', () => {
      mockConsole()

      const jcodNode = {
        key: 'any key',
        component: 'React',
        children: 'something',
        props: {
          className: 'any-class',
        },
        renderProps: {
          children: 'other something',
        },
      }
      const response = `The children of component [${jcodNode.component} // in position ${metadata.path}] have several values defined in [ROOT] and [renderProps]. The value of the children used is the one defined in [ROOT]. Other values are ignored.`

      ContentParser.showError(jcodNode, metadata)
      const spyError = console.error.mock.calls

      expect(Array.isArray(spyError)).toBe(true)
      expect(spyError).toEqual([[response]])
    })

    test('Should display 1 error: Children in ROOT & RenderProps & Props.', () => {
      mockConsole()

      const jcodNode = {
        key: 'any key',
        component: 'React',
        children: 'something',
        props: {
          children: 'other something',
          className: 'any-class',
        },
        renderProps: {
          children: 'other something again',
        },
      }
      const response = `The children of component [${jcodNode.component} // in position ${metadata.path}] have several values defined in [ROOT], [renderProps] and [props]. The value of the children used is the one defined in [ROOT]. Other values are ignored.`

      ContentParser.showError(jcodNode, metadata)
      const spyError = console.error.mock.calls

      expect(Array.isArray(spyError)).toBe(true)
      expect(spyError).toEqual([[response]])
    })

    test('Should display 1 error: Children in RenderProps & Props.', () => {
      mockConsole()

      const jcodNode = {
        key: 'any key',
        component: 'React',
        props: {
          children: 'something',
          className: 'any-class',
        },
        renderProps: {
          children: 'other something',
        },
      }
      const response = `The children of component [${jcodNode.component} // in position ${metadata.path}] have several values defined in [renderProps] and [props]. The value of the children used is the one defined in [renderProps]. Other values are ignored.`

      ContentParser.showError(jcodNode, metadata)
      const spyError = console.error.mock.calls

      expect(Array.isArray(spyError)).toBe(true)
      expect(spyError).toEqual([[response]])
    })

    test('Should NO display error: Children in RenderProps.', () => {
      mockConsole()

      const jcodNode = {
        key: 'any key',
        component: 'React',
        props: {
          className: 'any-class',
        },
        renderProps: {
          children: 'something',
        },
      }

      ContentParser.showError(jcodNode, metadata)
      const spyError = console.error.mock.calls

      expect(Array.isArray(spyError)).toBe(true)
      expect(spyError).toEqual([])
    })

    test('Should NO display error: Children in Props.', () => {
      mockConsole()

      const jcodNode = {
        key: 'any key',
        component: 'React',
        props: {
          children: 'something',
          className: 'any-class',
        },
      }

      ContentParser.showError(jcodNode, metadata)
      const spyError = console.error.mock.calls

      expect(Array.isArray(spyError)).toBe(true)
      expect(spyError).toEqual([])
    })

    test('Should NO display error: Children in ROOT.', () => {
      mockConsole()

      const jcodNode = {
        key: 'any key',
        component: 'React',
        children: 'something',
        props: {
          className: 'any-class',
        },
      }

      ContentParser.showError(jcodNode, metadata)
      const spyError = console.error.mock.calls

      expect(Array.isArray(spyError)).toBe(true)
      expect(spyError).toEqual([])
    })
  })

  describe('Test key value', () => {
    test('Should display 1 error: key in Props.', () => {
      mockConsole()

      const jcodNode = {
        component: 'React',
        children: 'something',
        props: {
          key: 'any key',
          className: 'any-class',
        },
      }
      const response = `The [props.key] with the value 'any key' of component [React // in position 0.0.1] is ignored. The key value, when defined, must be set in the ROOT of JCOD node.`

      ContentParser.showError(jcodNode, metadata)
      const spyError = console.error.mock.calls

      expect(Array.isArray(spyError)).toBe(true)
      expect(spyError).toEqual([[response]])
    })

    test('Should display 1 error: key in Props.', () => {
      mockConsole()

      const jcodNode = {
        component: 'React',
        children: 'something',
        props: {
          className: 'any-class',
        },
        renderProps: { key: 'any key' },
      }
      const response = `The [renderProps.key] with the value 'any key' of component [React // in position 0.0.1] is ignored. The key value, when defined, must be set in the ROOT of JCOD node.`

      ContentParser.showError(jcodNode, metadata)
      const spyError = console.error.mock.calls

      expect(Array.isArray(spyError)).toBe(true)
      expect(spyError).toEqual([[response]])
    })
  })

  describe('Test Multiple Error (Children & key)', () => {
    test('Should display 3 errors. (children, props.key, renderProps.key)', () => {
      const restoreConsole = mockConsole()

      const jcodNode = {
        component: 'React',
        children: 'something',
        props: {
          key: 'any key',
          children: 'other something',
          className: 'any-class',
        },
        renderProps: {
          key: 'any other key',
          children: 'other something again',
        },
      }

      ContentParser.showError(jcodNode, metadata)
      const spyError = console.error.mock.calls

      expect(Array.isArray(spyError)).toBe(true)
      expect(Array.isArray(spyError[0])).toBe(true)
      expect(spyError[0].length).toBe(3)

      restoreConsole()
    })
  })
})

// getComponent
describe('[ContentParser.getComponent] Identify good & falsy component.', () => {
  const availableComponent = {}
  // TODO : test On Error message
  // TODO : test for HTML & Custom-Element
  // TODO : test for Limited list of elment

  test('Should return a FalsyComponent (html element)', () => {
    const options = {}

    const ReturnedComponent = ContentParser.getComponent(
      'html',
      availableComponent,
      options,
    )

    expect(typeof ReturnedComponent).toBe('function')
    expect(ReturnedComponent.name).toBe('FalsyComponent')
  })

  test('Should return a HTML Element', () => {
    const options = { allowUnsecureHtmlElement: true }
    const element = 'html'

    const ReturnedComponent = ContentParser.getComponent(
      element,
      availableComponent,
      options,
    )

    expect(typeof ReturnedComponent).toBe('string')
    expect(ReturnedComponent).toBe(element)
  })

  test('Should return a FalsyComponent (custom-element)', () => {
    const options = {}

    const ReturnedComponent = ContentParser.getComponent(
      'custom-element',
      availableComponent,
      options,
    )

    expect(typeof ReturnedComponent).toBe('function')
    expect(ReturnedComponent.name).toBe('FalsyComponent')
  })

  test('Should return a custom-element', () => {
    const options = { allowUnsecureCustomElement: true }
    const element = 'custom-element'

    const ReturnedComponent = ContentParser.getComponent(
      element,
      availableComponent,
      options,
    )

    expect(typeof ReturnedComponent).toBe('string')
    expect(ReturnedComponent).toBe(element)
  })

  test('Should return a FalsyComponent (ReactComponent)', () => {
    const options = {}

    const ReturnedComponent = ContentParser.getComponent(
      'UnknowComponent',
      availableComponent,
      options,
    )

    expect(typeof ReturnedComponent).toBe('function')
    expect(ReturnedComponent.name).toBe('FalsyComponent')
  })

  test('Should return a ReactComponent', () => {
    const options = {}
    const ReactComponent = () => null
    const availableComponent = { ReactComponent }

    const ReturnedComponent = ContentParser.getComponent(
      ReactComponent.name,
      availableComponent,
      options,
    )

    expect(typeof ReturnedComponent).toBe('function')
    expect(ReturnedComponent.name).toBe(ReactComponent.name)
  })
})

describe('[ContentParser] Ennable the parser & render of JCOD tree.', () => {
  const JcodParser = ContentParser
  const jcodHtml = { component: 'div' }
  const jcodHtmlExtend = {
    component: 'div',
    children: 'Test Children',
    props: { anyProps: 'something as props value', className: 'class' },
  }
  const jcodCutomElement = { component: 'custom-element' }
  const jcodCutomElementExtend = {
    component: 'custom-element',
    children: 'Test Children',
    props: { anyProps: 'something as props value', className: 'class' },
  }
  const jcodReactComponent = { component: 'ReactComponent' }
  const jcodReactComponentExtend = {
    component: 'ReactComponent',
    children: 'Test Children',
    props: { anyProps: 'something as props value', className: 'class' },
  }
  const jcodTree = [
    'string node',
    jcodHtml,
    jcodHtmlExtend,
    jcodCutomElement,
    jcodCutomElementExtend,
    jcodReactComponent,
    jcodReactComponentExtend,
  ]

  const ReactComponent = ({children, ...props}) => <div {...props}>{children}</div>
  const availableComponent = {
    ReactComponent 
  }

  test('Should render only String node of a JCOD tree without crash', () => {
    const testRenderer = renderer.create(
      <JcodParser
        components={{}}
        data={jcodTree}
        options={{}}
      />,
    )
    console.log(testRenderer.toJSON())
    let tree = testRenderer.toJSON()
    expect(tree).toMatchSnapshot()
    expect(tree).toEqual('string node')
  })

  test('Should render only ReactComponent of a JCOD tree without crash', () => {
    const testRenderer = renderer.create(
      <JcodParser
        components={availableComponent}
        data={jcodTree}
        options={{}}
      />,
    )
    console.log(testRenderer.toJSON())
    let tree = testRenderer.toJSON()
    expect(tree).toMatchSnapshot()
    // expect(tree).toEqual('')
  })

  test('Should render ReactComponent and HTML element a simple JCOD tree without crash', () => {
    const testRenderer = renderer.create(
      <JcodParser
        components={availableComponent}
        data={jcodTree}
        options={{ allowUnsecureHtmlElement: true }}
      />,
    )
    console.log(testRenderer.toJSON())
    let tree = testRenderer.toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('Should render ReactComponent and custom-element a simple JCOD tree without crash', () => {
    const testRenderer = renderer.create(
      <JcodParser
        components={availableComponent}
        data={jcodTree}
        options={{ allowUnsecureCustomElement: true }}
      />,
    )
    console.log(testRenderer.toJSON())
    let tree = testRenderer.toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('Should render ReactComponent, HTML element and custom-element a simple JCOD tree without crash', () => {
    const testRenderer = renderer.create(
      <JcodParser
        components={availableComponent}
        data={jcodTree}
        options={{ allowUnsecureElements: true }}
      />,
    )
    console.log(testRenderer.toJSON())
    let tree = testRenderer.toJSON()
    expect(tree).toMatchSnapshot()
  })

  // constructor - After mount : Enable renderChild as methode of instance

  // render - Return the React Tree
})
