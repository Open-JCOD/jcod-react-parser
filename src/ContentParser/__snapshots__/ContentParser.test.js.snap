// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[ContentParser] Ennable the parser & render of JCOD tree. Should render ReactComponent and HTML element a simple JCOD tree without crash 1`] = `
Array [
  "string node",
  <div />,
  <div
    anyProps="something as props value"
    className="class"
  >
    Test Children
  </div>,
  <div />,
  <div
    anyProps="something as props value"
    className="class"
  >
    Test Children
  </div>,
]
`;

exports[`[ContentParser] Ennable the parser & render of JCOD tree. Should render ReactComponent and custom-element a simple JCOD tree without crash 1`] = `
Array [
  "string node",
  <custom-element />,
  <custom-element
    anyProps="something as props value"
    className="class"
  >
    Test Children
  </custom-element>,
  <div />,
  <div
    anyProps="something as props value"
    className="class"
  >
    Test Children
  </div>,
]
`;

exports[`[ContentParser] Ennable the parser & render of JCOD tree. Should render ReactComponent, HTML element and custom-element a simple JCOD tree without crash 1`] = `
Array [
  "string node",
  <div />,
  <div
    anyProps="something as props value"
    className="class"
  >
    Test Children
  </div>,
  <custom-element />,
  <custom-element
    anyProps="something as props value"
    className="class"
  >
    Test Children
  </custom-element>,
  <div />,
  <div
    anyProps="something as props value"
    className="class"
  >
    Test Children
  </div>,
]
`;

exports[`[ContentParser] Ennable the parser & render of JCOD tree. Should render only ReactComponent of a JCOD tree without crash 1`] = `
Array [
  "string node",
  <div />,
  <div
    anyProps="something as props value"
    className="class"
  >
    Test Children
  </div>,
]
`;

exports[`[ContentParser] Ennable the parser & render of JCOD tree. Should render only String node of a JCOD tree without crash 1`] = `"string node"`;
